pageserver/src/pgdatadir_mapping.rs

```diff
@@ -902,6 +902,56 @@ impl<'a, R: Repository> DatadirModification<'a, R> {
         Ok(())
     }
 
+    ///
+    /// Flush changes accumulated so far to the underlying repository.
+    ///
+    /// Usually, changes made in DatadirModification are atomic, but this allows
+    /// you to flush them to the underlying repository before the final `commit`.
+    /// That allows to free up the memory used to hold the pending changes.
+    ///
+    /// Currently only used during bulk import of a data directory. In that
+    /// context, breaking the atomicity is OK. If the import is interrupted, the
+    /// whole import fails and the timeline will be deleted anyway. (XXX: is it
+    /// actually deleted currently?)
+    ///
+    /// Note: A consequence of flushing the pending operations is that they
+    /// won't be visible to subsequent operations until `commit`. The function
+    /// retains all the metadata, but data pages are flushed. That's again OK
+    /// for bulk import, where you are just loading data pages and won't try to
+    /// modify the same pages twice.
+    pub fn flush(&mut self) -> Result<()> {
+        // Unless we have accumulated a decent amount of changes, it's not worth it
+        // to scan through the pending_updates list.
+        let pending_nblocks = self.pending_nblocks;
+        if pending_nblocks < 10000 {
+            return Ok(());
+        }
+
+        let writer = self.tline.tline.writer();
+
+        // Flush relation data blocks, keep metadata.
+        let mut result: Result<()> = Ok(());
+        self.pending_updates.retain(|&key, value| {
+            if result.is_ok() && is_rel_block_key(key) {
+                result = writer.put(key, self.lsn, value.clone());
```

Comment:
`value.clone()` seems to be redundant for a value that is removed anyway. However, I don't see any better ways to fix this. Maybe [`Vec::drain_filter`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain_filter), but that's nightly feature

Reply:
It seems that the end consumer has no need in the owned value. It is used just for serialization, so we can use a reference instead. See https://github.com/neondatabase/neon/compare/bulk-import...dkr/timeline-writer-value-reference?expand=1#diff-4f06906c7b86c8ddda375b63187440dc26a97298d1ddb431b27fc9a49d19aaa7R276

Reply:
Thanks! I added that to this PR